source/tests/integration/ni_fake_non_ivi_service_tests_endtoend.cpp

```diff
@@ -0,0 +1,232 @@
+
+namespace ni {
+namespace tests {
+namespace integration {
+
+using namespace nifake_non_ivi_grpc;
+
+// Test NiFakeNonIviService using a client stub.
+class NiFakeNonIviServiceTests_EndToEnd : public ::testing::Test {
+ public:
+  using FakeResourceRepository = nidevice_grpc::SessionResourceRepository<FakeHandle>;
+  nidevice_grpc::SessionRepository session_repository_;
+  std::shared_ptr<FakeResourceRepository> resource_repository_;
+  ni::tests::unit::NiFakeNonIviMockLibrary library_;
+  NiFakeNonIviService service_;
+  std::unique_ptr<::grpc::Server> server_;
+  std::unique_ptr<NiFakeNonIvi::Stub> stub_;
+  std::atomic<bool> shutdown_{false};
+
+  NiFakeNonIviServiceTests_EndToEnd()
+      : session_repository_(),
+        resource_repository_(std::make_shared<FakeResourceRepository>(&session_repository_)),
+        library_(),
+        service_(&library_, resource_repository_),
+        server_(start_server()),
+        stub_(NiFakeNonIvi::NewStub(server_->InProcessChannel(::grpc::ChannelArguments())))
+  {
+  }
+
+  virtual ~NiFakeNonIviServiceTests_EndToEnd()
+  {
+    shutdown();
+  }
+
+  void shutdown()
+  {
+    if (!shutdown_) {
+      shutdown_ = true;
+      server_->Shutdown();
+    }
+  }
+
+  std::unique_ptr<::grpc::Server> start_server()
+  {
+    ::grpc::ServerBuilder builder;
+    builder.RegisterService(&service_);
+    return builder.BuildAndStart();
+  }
+
+  std::unique_ptr<NiFakeNonIvi::Stub>& stub()
+  {
+    return stub_;
+  }
+};
+
+TEST_F(NiFakeNonIviServiceTests_EndToEnd, ReadStream_ReturnsExpectedSequence)
+{
+  ::grpc::ClientContext context;
+  const auto START = 0;
+  const auto STOP = 100;
+  ReadStreamRequest request;
+  request.set_start(START);
+  request.set_stop(STOP);
+  auto reader = stub()->ReadStream(&context, request);
+
+  ReadStreamResponse response;
+  for (auto i = START; i <= STOP; ++i) {
+    reader->Read(&response);
+    EXPECT_EQ(response.value(), i);
+  }
+}
+
+TEST_F(NiFakeNonIviServiceTests_EndToEnd, ReadPartialStream_ReturnsExpectedSequence)
+{
+  ::grpc::ClientContext context;
+  const auto START = 0;
+  const auto STOP = 100;
+  ReadStreamRequest request;
+  request.set_start(START);
+  request.set_stop(STOP);
+  auto reader = stub()->ReadStream(&context, request);
+
+  ReadStreamResponse response;
+  for (auto i = START; i <= STOP / 2; ++i) {
+    reader->Read(&response);
+    EXPECT_EQ(response.value(), i);
+  }
+}
+
+TEST_F(NiFakeNonIviServiceTests_EndToEnd, ReadTwoStreamsConsecutively_ReturnsExpectedSequences)
+{
+  ::grpc::ClientContext context;
+  const auto START = 0;
+  const auto STOP = 100;
+  const auto OFFSET = 100;
+  ReadStreamRequest request;
+  request.set_start(START);
+  request.set_stop(STOP);
+  auto reader = stub()->ReadStream(&context, request);
+  request.set_start(START + OFFSET);
+  request.set_stop(STOP + OFFSET);
+  ::grpc::ClientContext second_context;
+  auto second_reader = stub()->ReadStream(&second_context, request);
+  reader->WaitForInitialMetadata();
```

Comment:
Do we need to call `WaitForInitialMetadata`? I don't see us explicitly sending initial metadata from this ni fake service handler unless it's done behind the scenes somewhere. And if it is needed, for things with two stream do we need to call `WaitForInitialMetadata` for both readers?

Reply:
DAQ cares about `WaitForInitialMetadata` because it requires that you set register before calling `DAQmxStartTask`. This test doesn't have an ordering dependency like that. Also, because it starts updating immediately, it doesn't really need a SendInitialMetadata because the first write will do that.

I focused more on stress testing the stream/lifetime code than trying to copy the DAQ case.

Reply:
Yeah, I understood it for the DAQ case and was curious if we needed to care/know about it for these tests. That makes sense though and is more clear. It seems fine as is. I'll mark this resolved.